src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -3931,6 +3958,60 @@ void verify(VarPatternSyntax declaration, string name, string expectedType)
         }
     }
 
+    [Fact]
+    public void SlicePattern_Nullability_Annotation()
+    {
+        var source = @"
+#nullable enable
+class C
+{
+    public int Length => throw null!;
+    public int this[int i] => throw null!;
+    public int[]? Slice(int i, int j) => throw null!;
+
+    public void M()
+    {
+        if (this is [1, ..var slice])
+            slice.ToString();
+        if (this is [1, ..[] list])
+            list.ToString();
+    }
+}
+";
+        var compilation = CreateCompilationWithIndexAndRange(source);
+        compilation.VerifyEmitDiagnostics(
+            // (12,13): warning CS8602: Dereference of a possibly null reference.
+            //             slice.ToString();
+            Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "slice").WithLocation(12, 13),
+            // (14,13): warning CS8602: Dereference of a possibly null reference.
+            //             list.ToString();
+            Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "list").WithLocation(14, 13)
```

Comment:
I'm not sure I understand this one. Shouldn't the `[]` pattern say that it's a non-null slice, since it has to dereference the slice to determine if the slice is empty?

Reply:
Makes sense. I'd defer to @jcouv for that. My solution was to ignore and report the annotation on the return type.

If we consider slice.Length evaluation (or any other) as a dereference with no warning on forthcoming derefs, shouldn't we inform the user about it?

Reply:
I think we can do something in between: ignore and report only if there's a dereference i.e. when the result is being used - that would exclude var patterns.

Reply:
Thanks for spotting this. There's indeed a good question here.
From the LDM notes: "We will adjust our expectations around slice patterns to assume that Slice will never return a null value, and we will not insert a null test into the reachability or codegen DAGs."

I think it should follow that both the `.. var slice` and `.. [] slice2` cases should yield not-null state for those variables.

Reply:
@jcouv Can you please take another look? thanks.